### PR TITLE
New header refactoring

### PR DESCRIPTION
**Pull request by gu-stav**
## What does this change?

This is an attempt to collect all the CSS, which was added in various places during the new header redesign and have it at a central place, with a modular structure.

### Principles

- Parents control the relation between it's children (Spacing, ...)
- As soon as something needs a modifier, it becomes a module
- Things like `navigation`, `subnav` or `fly-in` don't know anything about their content. They only handle the relations between the children. Therefore `fly-in` contains for example groups or search-widgets, instead of defining them as part of the module ...

### Known ~issues~ challenges

- Naming: it's always hard. Right now everything, except `subnav` is nested within `new-header/` which makes clear, those modules are only useful for the module itself. Another, probably better way would be to namespace them to `new-header-fly-in` for example.

## What is the value of this and can you measure success?

Header becomes a module, readability, less duplication and developer happiness.

## Does this affect other platforms - Amp, Apps, etc?

Amp.

## Screenshots

soon.

## Tested in CODE?

Not yet.

### CODE REVIEW

File: static/src/stylesheets/amp/_header.scss [hunk 1]
```diff
@@ -1,334 +1,16 @@
...
-        // Override a from user agent stylesheet
-        color: inherit;
-    }
-}
-
-.secondary-navigation {
-    font-size: .9em;
-    margin: 0;
-}
+@import '../layout/new-header/_variables';
```
Comment (NataliaLKB) [inline static/src/stylesheets/amp/_header.scss]: Purposefully, the css for the new header on AMP is lighter.

I know the (sort of) duplication is annoying, but because of the AMP restrictions on css its best not to just import the same files - a lot of them contain css we don't need in AMP. 

So this PR doesn't get blocked (and we can continue to work on the new header) I would take all the AMP changes out of this PR, or at least try and break up this refactoring in much smaller PRs.

Its a lot to review as is.

**Comment by NataliaLKB** [comment]
FYI @zeftilldeath @stephanfowler

File: static/src/stylesheets/amp/_header.scss (hunk 1 above)
Comment (gustavpursche) [inline static/src/stylesheets/amp/_header.scss]: I think we can share most of it - I'm trying to come up with something that takes the non-mobile related stuff out of the AMP include. All the rest is shareable I think.

Let's do not review this yet and go through it on Monday together. I think that's much easier.

Comment (NataliaLKB) [inline static/src/stylesheets/amp/_header.scss]: > I think we can share most of it - I'm trying to come up with something that takes the non-mobile related stuff out of the AMP include. All the rest is shareable I think.

Potentially, but we did simplify the design slightly for AMP as well

> Let's do not review this yet and go through it on Monday together. I think that's much easier.

Even then, this should be broken up into smaller changes.

**Comment by NataliaLKB** [comment]
I think it would be good in advance of Monday to write down all the reasons for this change as clearly as possible. Less duplication is obviously one reason, but there is a lot more that has changed other than that. Explaining all the problems you are trying to solve and how you have solved them with this PR would really help the review process. 

Refactoring things is really great, and I am so happy to you are doing it. A lot of this code is much better, but I worry the AMP changes will block this PR from going through as is.

As you said, lets sit down next week and go through the changes 💯

**Comment by gustavpursche** [comment]
@NataliaLKB @zeftilldeath I think this is ready for review now. ✨

**Comment by NataliaLKB** [comment]
Thanks for making those changes Gustav! Do you mind squashing these commits and rebasing with master? There seems to be conflicts.

File: article/test/ArticleFeatureTest.scala [hunk 2]
```diff
@@ -494,7 +494,6 @@ import collection.JavaConversions._
         findFirst("aside").getAttribute("role") should be("complementary")
         findFirst("header").getAttribute("role") should be("banner")
         findFirst(".l-footer__secondary").getAttribute("role") should be("contentinfo")
-        findFirst("nav").getAttribute("role") should be("navigation")
```
Comment (NataliaLKB) [inline article/test/ArticleFeatureTest.scala]: Why is this being deleted?

File: common/app/views/fragments/nav/editionPicker.scala.html [hunk 3]
```diff
@@ -3,24 +3,23 @@
 @import common.{Edition, LinkTo}
 @import views.support.RenderClasses
 
-<li class="navigation-group__item navigation-group__item--edition-picker">
+<li class="fly-in-item">
```
Comment (NataliaLKB) [inline common/app/views/fragments/nav/editionPicker.scala.html]: I am not sure I fully get this name. To me this won't work if we try to use the same markup for desktop. What about just calling it `menu` or something more generic?

File: article/test/ArticleFeatureTest.scala (hunk 2 above)
Comment (gustavpursche) [inline article/test/ArticleFeatureTest.scala]: Because `<nav>` has the role baked into it's semantics. There is no meaning in duplicating the logic as an `aria-role` except for browsers who would treat `nav` as `div`. I think we don't support any of them anymore.

Comment (NataliaLKB) [inline article/test/ArticleFeatureTest.scala]: When we say we don't support those browsers, it means we don't care if they look bad. They still have to usable. I think the guidance is that content needs to be readable/viewable and you need to be able to navigate. 

The navigation definitely needs to at least sort of work for browsers we don't support.

We say:
> We believe that our journalism should be accessible to everyone.

and 

> This means every feature of the website needs to consider how it adapts itself to each user's environment.

in our [Browser Support Principals](https://github.com/guardian/frontend/blob/master/docs/04-quality/02-browser-support-principles.md#browser-support-principles). But I think we should make this more clear and probably give examples of what this means in practise

Comment (gustavpursche) [inline article/test/ArticleFeatureTest.scala]: Well - we are talking basically about IE8 here. The page is completly broken there already (no styles, no javascript).

IE8:

screen shot 2017-05-25 at 09 55 20

`nav` support matrix:

screen shot 2017-05-25 at 09 53 06

Adding the `aria-role` for these browser is an enhancement on top. It has nothing to do with the functionality, so I think it's ok to not add it anymore.

Comment (NataliaLKB) [inline article/test/ArticleFeatureTest.scala]: Shame that it's so broken now on IE8. It didn't used to be! I wonder how it ended up so broken within the last 6 months. Could it be the the introduction of webpack? 

Not saying we should now fix it, but it is a shame as 0.34% of the world still uses that browser. Thats a lot of people we can't reach at all now. Hopefully next year no one will be using it anymore.

File: common/app/views/fragments/amp/header.scala.html [hunk 4]
```diff
@@ -3,12 +3,18 @@
...
 
-<header class="header" role="banner">
+<header class="header"
+        role="banner">
     <div class="header__inner gs-container">
-        <a class="header__supporter-cta" data-link-name="amp : nav : supporter-cta"
-        href="@Configuration.id.membershipUrl/supporter?INTCMP=AMP_HEADER_GU_SUPPORTER">
-            Become a <span>Supporter</span>
-        </a>
+        <div class="header__support">
```
Comment (NataliaLKB) [inline common/app/views/fragments/amp/header.scala.html]: I think this element is a `cta` (call to action) and isn't really a "support" of of the header. Maybe rename to: `header__cta`?

File: static/src/javascripts/projects/common/modules/navigation/newHeaderNavigation.js [hunk 5]
```diff
@@ -6,8 +6,7 @@ import userAccount from 'common/modules/navigation/user-account';
 
 const enhanced = {};
 
-const getSidebarElement = (): ?HTMLElement =>
-    document.getElementById('main-menu');
+const getFlyIn = (): ?HTMLElement => document.getElementById('main-menu');
```
Comment (NataliaLKB) [inline static/src/javascripts/projects/common/modules/navigation/newHeaderNavigation.js]: Still not so sure about this name. Without you explaining it I can't visualise what "fly in" could mean. Event `flyInMenu` would help. Or what about `mainMenu` as that is what it is , no?

File: static/src/stylesheets/amp/_header.scss [hunk 6]
```diff
@@ -5,6 +5,8 @@ $gutter-large: 55px;
 $veggie-burger-small: 42px;
 $veggie-burger-medium: 52px;
```
Comment (NataliaLKB) [inline static/src/stylesheets/amp/_header.scss]: Should we change this value to `54px` as well?

File: static/src/stylesheets/layout/new-header/_menu-group.scss [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+.menu-group--highlight {
+    color: $guardian-brand-dark;
+    background-color: $news-main-2;
+
+    .menu-item:last-child .menu-group {
+        margin-bottom: -$gs-baseline;
+    }
+}
+
+.menu-group--downlight {
```
Comment (NataliaLKB) [inline static/src/stylesheets/layout/new-header/_menu-group.scss]: Like the naming and organisation here! Awesome 💯

File: static/src/stylesheets/layout/new-header/_menu-item.scss [hunk 8]
```diff
@@ -0,0 +1,127 @@
+.menu-item {
+    position: relative;
+    width: 100%;
+
+    > *[data-link-name='nav2 : the guardian app'],
```
Comment (NataliaLKB) [inline static/src/stylesheets/layout/new-header/_menu-item.scss]: Do you think we should have classnames here instead?

Comment (NataliaLKB) [inline static/src/stylesheets/layout/new-header/_menu-item.scss]: (I know this was already there, but just a question)

File: static/src/stylesheets/layout/new-header/_menu-item.scss [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+    }
+}
+
+.menu-item__title {
+    color: currentColor;
+    cursor: pointer;
+    display: block;
+    font-size: 18px;
+    outline: none;
+    padding-bottom: $gs-baseline / 1.8;
```
Comment (NataliaLKB) [inline static/src/stylesheets/layout/new-header/_menu-item.scss]: `$gs-baseline / 1.8` ?? Surely this should be just `$gs-baseline / 2`. I don't think we should promote things that go against our grid system that much

File: static/src/stylesheets/layout/new-header/_menu-item.scss [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+    padding-right: $gs-gutter;
+    padding-top: $gs-baseline / 1.8;
+    user-select: none;
+
+    &:hover,
+    &:focus {
+        text-decoration: none;
+    }
+
+    // only for Chrome/ Safari - FF loses the marker via display: block
```
Comment (NataliaLKB) [inline static/src/stylesheets/layout/new-header/_menu-item.scss]: Really useful comment!

File: static/src/stylesheets/layout/new-header/_menu-item.scss [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+    }
+
+    .menu-group--primary & {
+        font-size: 20px;
+        padding-bottom: $gs-baseline * 1.2;
+        padding-top: $gs-baseline / 2.5;
+
+        @include mq(mobileMedium) {
+            font-size: 22px;
+            padding-top: $gs-baseline / 2.3;
```
Comment (NataliaLKB) [inline static/src/stylesheets/layout/new-header/_menu-item.scss]: Again, not so sure about all these `.x` division values. Chat to @zeftilldeath to see if we can better align them with our grid system or else just use a `px` value so at least its less confusing to read.

File: static/src/stylesheets/layout/new-header/_menu-search.scss [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+    position: relative;
+}
+
+.menu-search__search-box {
+    background-color: $guardian-brand;
+    border: 0;
+    border-radius: 1000px;
+    box-sizing: border-box;
+    font-size: 16px;
+    height: 36px;
```
Comment (NataliaLKB) [inline static/src/stylesheets/layout/new-header/_menu-search.scss]: We can use `$gs-row-height` here https://github.com/guardian/guss-grid-system

File: static/src/stylesheets/layout/new-header/_new-header.scss [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+    @include mq(mobileLandscape) {
+        left: $gs-gutter / 2;
+    }
+}
+
+.new-header__logo {
+    margin-left: auto;
+}
+
+.new-header__logo .inline-logo__svg {
```
Comment (NataliaLKB) [inline static/src/stylesheets/layout/new-header/_new-header.scss]: Add a custom class name so we can target the svg without nesting?

File: static/src/stylesheets/layout/new-header/_support-item.scss [hunk 14]
```diff
@@ -1,58 +1,52 @@
...
         }
     }
+}
 
-    @include mq($from: mobileLandscape) {
-        left: $gs-gutter/2;
+.support-item__label {
+    color: $guardian-brand-dark;
+    display: block;
+    font-family: 'Guardian Egyptian Web', 'Guardian Text Egyptian Web', Georgia, serif;
```
Comment (NataliaLKB) [inline static/src/stylesheets/layout/new-header/_support-item.scss]: Can we use a font variable here, or a mixin is preferred.

File: static/src/stylesheets/module/facia/_mixins.scss [hunk 15]
```diff
@@ -22,15 +22,15 @@ $breakpoints: (
             margin-left: $gs-gutter;
             margin-right: $gs-gutter;
         }
-        @include mq(containerWidestMobile) {
```
Comment (NataliaLKB) [inline static/src/stylesheets/module/facia/_mixins.scss]: How are these changes relevant for this change? 

Given its a facia mixin I just want to confirm that it's changing something unexpected.

File: static/src/stylesheets/layout/new-header/_new-header.scss (hunk 13 above)
Comment (gustavpursche) [inline static/src/stylesheets/layout/new-header/_new-header.scss]: Yes, will do. I left it that way, because the inline functions breaks the BEM naming system (produces `.header__logo__svg`), but since it's all over the place I guess you are right.

File: static/src/stylesheets/module/facia/_mixins.scss (hunk 15 above)
Comment (gustavpursche) [inline static/src/stylesheets/module/facia/_mixins.scss]: This as agreed with Stephan and Zef. I'll double check, when Zef is doing his review ...

Comment (NataliaLKB) [inline static/src/stylesheets/module/facia/_mixins.scss]: Cool! No sense changing it now, but for future probably best to move this to a separate PR so the change can be properly documented in isolation (I have a vague memory of us discussing it, but if I didn't have that memory there would be no documented reason why this change happened)

File: static/src/stylesheets/layout/new-header/_menu-item.scss (hunk 8 above)
Comment (gustavpursche) [inline static/src/stylesheets/layout/new-header/_menu-item.scss]: My first reflex about adding classnames was the same, but then I thought it's actually a good way of saying: hey - something special is going on here. If we don't do it here, we'd have to hide the ugly logic for doing this in Scala and so, in the sense of "move complexity to the edge" I'm in a favor of keeping it. Also: we won't see this in the desktop design, so the classes would be useless there ...

Comment (NataliaLKB) [inline static/src/stylesheets/layout/new-header/_menu-item.scss]: All good points!

**Review by NataliaLKB — APPROVED** [review]
There are a couple outstanding comments which @zeftilldeath can probably help with. Other than those it looks great!
